### PR TITLE
AzureMonitor: Managed Identity configuration UI

### PR DESCRIPTION
**Pull request by kostrse**
**What this PR does / why we need it**:

Switching between **managed identity** and **app registration** (client secret) authentication types in the Azure Monitor configuration editor. This new logic takes advantage of the new [Azure token provider](https://github.com/grafana/grafana/blob/main/pkg/api/pluginproxy/token_provider_azure.go) which was recently implemented at the Grafana server.

Managed Identity authentication is only available as a choice in the UI only if it was explicitly enabled in [Grafana config](https://github.com/grafana/grafana/blob/65cacd31c2f4089bc644d0d6f86372c8cadd5c57/conf/defaults.ini#L559).

**Which issue(s) this PR fixes**:

Continuation of the work outlined in #33721 

**Special notes for your reviewer**:

This is draft PR, there are still use cases that need to be fixed.

### CODE REVIEW

**Comment by kostrse** [comment]
@joshhunt @sarahzinger the configuration UI is ready for review

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts
Comment (sarahzinger) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts:503]: it strikes me as a bit odd that a truthy result from isValidConfig is actually an invalid config but it looks like we already have that naming convention haha. I also think explicitly returning undefined is sort of strange, but we also already do that so I suppose this is probably not worth changing at the moment. Perhaps something todo later.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/AnalyticsConfig.tsx
Comment (sarahzinger) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/AnalyticsConfig.tsx:30]: it seems like we manually check against the strings 'clientsecret' and 'msi' a lot. Seems like it might be worthwhile to turn these into typescript enums so you could do something here like `primaryCredentials.authType === supportedAuthTypes.clientSecret` or something like that?

**Review by sarahzinger — COMMENTED** [review]
Would be nice to have someone with more context into this give another set of eyes, but I have some ideas/suggestions to think about.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts
Comment (joshhunt) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts:503]: hmmm I yeah @kostrse I understand this is the same pattern as is elsewhere (unfortunately), but you do you think we could rename the function (just this one here) just to make it a bit clearer, to reduce confusion/temptation with `if (isValidConfig()) return success`?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/credentials.ts
Comment (joshhunt) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/credentials.ts:27]: Excellent comments, very clear! 👍

**Review by joshhunt — COMMENTED** [review]
Looks good from me, but I'll leave the approval to @sunker as he's been a lot more familiar with this work.

Awesome stuff though!

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts
Comment (kostrse) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts]: Yes, this is existing implementation and it has even deeper problems than just naming :) There's  [isConfigured](https://github.com/grafana/grafana/blob/4779de9467857b80658574ace2f41ff14bcd25a5/public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts#L69-L71) method which is also does validation called in datasource.ts [here](https://github.com/grafana/grafana/blob/4779de9467857b80658574ace2f41ff14bcd25a5/public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts#L145-L155) even before the `isValidConfig` has chance to return validation results. So, in most cases error messages from `isValidConfig` never even returned.

I made some consmetic chagnes, I renamed the method to `validateDatasource`, added return type etc. But validation needs much deeper overhaul.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/AnalyticsConfig.tsx
Comment (kostrse) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/AnalyticsConfig.tsx:30]: `AzureAuthType` is actually [strongly-typed](https://github.com/grafana/grafana/blob/4779de9467857b80658574ace2f41ff14bcd25a5/public/app/plugins/datasource/grafana-azure-monitor-datasource/types/index.ts#L50), compiler will even return error if you try to add a new value and won't update the code which uses switch statements etc.

**Review by sunker — COMMENTED** [review]
Looks great @kostrse! Just need to test this on a VM in azure before I can approve. 

Question: If you attempt to use Managed Identity and auth fails for some reason, you're not able to load subscriptions. I think it would be nice to show some sort of error message in the ConfigEditor to indicate to the user that it didn't work. At the moment, you need to inspect the network traffic to know that something failed. 

Also, did we talk about disabling Azure Application Insights in the ConfigEditor in case you're creating a new data source or if you did not already have that configured? I can't remember? @joshhunt?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/AzureCredentialsForm.tsx
Comment (kostrse) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/AzureCredentialsForm.tsx:97]: The callback won't be called for anything over than `clientsecret` because otherwise controls are hidden but this check needed for type safety, to cast to the type. `credentials` is of generic type, so we should cast to `clientsecret` to be able to access the fields

**Comment by kostrse** [comment]
> Question: If you attempt to use Managed Identity and auth fails for some reason, you're not able to load subscriptions. I think it would be nice to show some sort of error message in the ConfigEditor to indicate to the user that it didn't work. At the moment, you need to inspect the network traffic to know that something failed.

Yes, right now either msi or clientsecret, if config editor is unable to fetch subscriptions, drop down list just silently doesn't update. This is existing behavior.

I didn't cover changes to error handing in this PR.

Created issue for this: #34426

> Also, did we talk about disabling Azure Application Insights in the ConfigEditor in case you're creating a new data source or if you did not already have that configured? I can't remember? @joshhunt?

We have issue for this: #33709

Also one more issue for datasource validation improvements: #34431

**Review by sunker — APPROVED** [review]
Tested on a VM in Azure and it works as expected. Fantastic work!

**Pull request merged**
